pkg/auth/iam/iam.go

```diff
@@ -90,57 +90,81 @@ func (filter *Filter) Auth(opts ...FilterOption) restful.FilterFunction {
 			return
 		}
 
-		claims, err := filter.iamClient.ValidateAndParseClaims(token)
-		if err != nil {
-			logrus.Warn("unauthorized access: ", err)
-			logIfErr(resp.WriteHeaderAndJson(http.StatusUnauthorized, ErrorResponse{
-				ErrorCode:    UnauthorizedAccess,
-				ErrorMessage: ErrorCodeMapping[UnauthorizedAccess],
-			}, restful.MIME_JSON))
+		filter.validateUser(opts, req, resp, chain, token, tokenFrom)
+	}
+}
 
+// PublicAuth returns a filter that allow unauthenticate request and request with valid access token in auth header or cookie
+// If request has acces token, the token's claims will be passed in the request.attributes["JWTClaims"] = *iam.JWTClaims{}
+// This filter is expandable through FilterOption parameter
+// Example:
+// iam.PublicAuth(
+// 		WithValidUser(),
+//		WithPermission("ADMIN"),
+// )
+func (filter *Filter) PublicAuth(opts ...FilterOption) restful.FilterFunction {
+	return func(req *restful.Request, resp *restful.Response, chain *restful.FilterChain) {
+		token, tokenFrom, err := parseAccessToken(req)
+		if err != nil {
+			chain.ProcessFilter(req, resp)
 			return
 		}
 
-		req.SetAttribute(ClaimsAttribute, claims)
+		filter.validateUser(opts, req, resp, chain, token, tokenFrom)
```

Comment:
the behavior expected for this filter is allow user without access token also allow valid user with specific authentication. this behavior is needed because we want to allow public user to get general data, but authenticated user will get additional data related to their user, in example for like count, creator following, etc

Reply:
Is it better to have a separate function to check whether the token is exists or not in other function instead of we leverage the parseAccessToken() result? We need to check from the header and the cookies in this case

Reply:
@ardhiyanyanu but from the implementation, take an example, if the user present an expired token, it will throw error instead success but not authenticated?

Reply:
@angelo-chan already change behavior of this `publicAuth` filter, so invalid token or not authenticated will treated as public. this will make behavior more consistent